File: leetcode_75/Binary_Tree_DFS/cpp/1372.cpp
#include <algorithm>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int longestZigZag(TreeNode *root)
    {
        return max(this->helper(root->right, 0, true), this->helper(root->left, 0, false));
    }

private:
    int helper(TreeNode *node, int count, bool to_right)
    {
        if (node == nullptr)
        {
            return count;
        }

        count++;
        if (to_right)
        {
            return max(this->helper(node->left, count, false), this->helper(node->right, 0, true));
        }
        else
        {
            return max(this->helper(node->right, count, true), this->helper(node->left, 0, false));
        }
    }
};

File: leetcode_75/Binary_Tree_DFS/python/1372.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def get_mx_path(self, node: TreeNode, count: int, last_action: str):
        if node is None:
            return count
        
        count += 1
        if last_action == "left":
            return max(self.get_mx_path(node.right, count, "right"), self.get_mx_path(node.left, 0, "left"))
        else:
            return max(self.get_mx_path(node.left, count, "left"), self.get_mx_path(node.right, 0, "right"))

    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        return max(self.get_mx_path(root.left, 0, "left"), self.get_mx_path(root.right, 0, "right"))
